### PR TITLE
[analyzer] Fix crash in BasicValueFactory.cpp with __int128_t integers

### PR DESCRIPTION
**Pull request by vabridgers**
This change avoids a crash in BasicValueFactory by checking the bit width of an APSInt to avoid calling getZExtValue if greater than 64-bits. This was caught by our internal, randomized test generator.

Clang invocation
clang -cc1 -analyzer-checker=optin.portability.UnixAPI case.c

<src-root>/llvm/include/llvm/ADT/APInt.h:1488:
uint64_t llvm::APInt::getZExtValue() const: Assertion `getActiveBits() <= 64
  && "Too many bits for uint64_t"' failed.
...

 #9 <address> llvm::APInt::getZExtValue() const
     <src-root>/llvm/include/llvm/ADT/APInt.h:1488:5
     clang::BinaryOperatorKind, llvm::APSInt const&, llvm::APSInt const&)
     <src-root>/clang/lib/StaticAnalyzer/Core/BasicValueFactory.cpp:307:37
     llvm::IntrusiveRefCntPtr<clang::ento::ProgramState const>,
     clang::BinaryOperatorKind, clang::ento::NonLoc, clang::ento::NonLoc,
     clang::QualType)
     <src-root>/clang/lib/StaticAnalyzer/Core/SimpleSValBuilder.cpp:531:31
     llvm::IntrusiveRefCntPtr<clang::ento::ProgramState const>,
     clang::BinaryOperatorKind, clang::ento::SVal, clang::ento::SVal,
     clang::QualType)
     <src-root>/clang/lib/StaticAnalyzer/Core/SValBuilder.cpp:532:26
...

### CODE REVIEW

File: clang/test/Analysis/int128-nocrash.c
Comment (DonatNagyE) [inline clang/test/Analysis/int128-nocrash.c:6]: Why did you add the "`| 1`"? Does the test work without it?

File: clang/lib/StaticAnalyzer/Core/BasicValueFactory.cpp
Comment (DonatNagyE) [inline clang/lib/StaticAnalyzer/Core/BasicValueFactory.cpp:278]:
```suggestion
      if (V2.isNegative() || V2.getBitWidth() > 64)
```
You can combine the new condition with the previous early return branch. Note that the signedness check was completely superfluous, because `APSInt::isNegative()` is defined as `isSigned() && APInt::isNegative()` [where `APInt::isNegative()` checks the sign bit].

File: clang/lib/StaticAnalyzer/Core/BasicValueFactory.cpp
Comment (DonatNagyE) [inline clang/lib/StaticAnalyzer/Core/BasicValueFactory.cpp:304]:
```suggestion
      if (V2.isNegative() || V2.getBitWidth() > 64)
```
As above.

**Review by NagyDonat — COMMENTED** [review]
Good commit! I was suspicious at first because `getBitWidth()` tends to be problematic, but I couldn't find any concrete case where the new code behaves incorrectly.

File: clang/test/Analysis/int128-nocrash.c
Comment (vabridgers) [reply clang/test/Analysis/int128-nocrash.c:6]: Hmm, I added this for no particular reason at the time. I'll try without it. I could not find a numeric literal suffix for a 128 bit number, this case was originally found by multiplying 2 64-bit numbers into a 128-bit number, and I simplified that.

File: clang/lib/StaticAnalyzer/Core/BasicValueFactory.cpp
Comment (vabridgers) [reply clang/lib/StaticAnalyzer/Core/BasicValueFactory.cpp:278]: Will do, thanks

File: clang/lib/StaticAnalyzer/Core/BasicValueFactory.cpp
Comment (vabridgers) [reply clang/lib/StaticAnalyzer/Core/BasicValueFactory.cpp:304]: Will do, thanks

**Comment by vabridgers** [comment]
I'll resolve the comments in an update. Thanks for the comments!

**Comment by vabridgers** [comment]
> I support changes like this. However, I think we should prefer reusing existing test files to creating more and more new ones. That has one benefit to me, one can see multiple cases of the topic on one screen without jumping around to open multiple. But that's all. I don't think I'll participate in the review though. They are piling up already.

Thanks @steakhal, I'll look at including this test into an existing case. Best

**Comment by vabridgers** [comment]
@steakhal, I looked at the existing test cases, but cannot find an existing test where this case would easily fit. I'm happy to follow a specific suggestion about where or how to reformat or rename this test. 

For now, I'll assume this current approach is ok until I see some concrete suggestions. 

Thanks

**Comment by vabridgers** [comment]
Hey guys, is it ok to merge this fix? Thanks

**Review by NagyDonat — APPROVED** [review]
It's simple and clean, let's just merge it :)

File: clang/test/Analysis/int128-nocrash.c
Comment (AaronBallman) [inline clang/test/Analysis/int128-nocrash.c:6]: It'd be good to add a test case using `_BitInt(72)` or something along those lines to ensure we handle that properly as well.

**Review by AaronBallman — COMMENTED** [review]
Please be sure to add a release note about the fix, too!

File: clang/test/Analysis/int128-nocrash.c
Comment (vabridgers) [reply clang/test/Analysis/int128-nocrash.c:6]: I added a test case. Let me know if you're ok with this. Thanks!

**Comment by vabridgers** [comment]
I'll wait for an ok from @AaronBallman before merging. Thank you!

File: llvm/docs/ReleaseNotes.rst
Comment (AaronBallman) [inline llvm/docs/ReleaseNotes.rst:218]: I think the release notes should go into `clang/docs/ReleaseNotes.rst` instead of LLVM's release notes. Also, be sure to add the markup to change links into actual links and formats syntax as such, along the lines of the suggested changes (you'll have to fix all the links up, I just did one as an example).

```suggestion
* A few crashes have been found and fixed using randomized testing related
  to the use of ``_BitInt()`` in tidy checks and in clang analysis. See
 `#67212 <https://github.com/llvm/llvm-project/pull/67212>`_,
  https://github.com/llvm/llvm-project/pull/66782,
  https://github.com/llvm/llvm-project/pull/65889,
  https://github.com/llvm/llvm-project/pull/65888, and
  https://github.com/llvm/llvm-project/pull/65887

```

Comment (vabridgers) [reply llvm/docs/ReleaseNotes.rst]: addressed

**Review by AaronBallman — APPROVED** [review]
Assuming precommit CI comes back green, the changes LGTM, thank you!

**Pull request merged**
